File: src/Java/Programmers/Lv2/PR_2019카카오개발자겨울인턴십_튜플.java
package Java.Programmers.Lv2;

import java.util.*;

class Tuple{
    public int[] tuple(String s){
        StringBuilder sb = new StringBuilder();
        HashMap<Integer, Integer> hm = new HashMap<>();
        boolean flag = false;
        for(int i = 0; i < s.length(); i++){
            char ch = s.charAt(i);
            if(ch >= '0' && ch <= '9'){
                sb.append(ch);
                flag = true;
            }else{
                if(flag){
                    int value = Integer.parseInt(sb.toString());
                    hm.put(value, hm.getOrDefault(value, 0) + 1);
                    sb = new StringBuilder();
                    flag = false;
                }
            }
        }

        int[] result = new int[hm.size()];
        int p = 0;
        List<Map.Entry<Integer, Integer>> entryList = new LinkedList<>(hm.entrySet());
        entryList.sort(new Comparator<Map.Entry<Integer, Integer>>() {
            @Override
            public int compare(Map.Entry<Integer, Integer> o1, Map.Entry<Integer, Integer> o2) {
                return o2.getValue() - o1.getValue();
            }
        });

        for(Map.Entry<Integer, Integer> entry : entryList){
            result[p] = entry.getKey();
            p++;
        }

        return result;
    }

    public static void main(String[] args){
        String s = "{{2},{2,1},{2,1,3},{2,1,3,4}}";

        Tuple sol = new Tuple();
        System.out.println(Arrays.toString(sol.tuple(s)));
    }
}

File: src/Kotlin/PR_2019카카오개발자겨울인턴십_튜플.kt
package Kotlin.tuple

import kotlin.collections.HashMap

class Solution{
    fun solution(s: String): IntArray {
        val hm = HashMap<Int, Int>()
        var sb = StringBuilder()
        var flag: Boolean = false
        for(i in 0 until s.length){
            val ch: Char = s[i]

            if(ch in '0'..'9'){
                sb.append(ch)
                flag = true
            }else{
                if(flag) {
                    val value: Int = sb.toString().toInt()
                    if(hm.containsKey(value)) hm[value] = hm[value]!! + 1
                    else hm[value] = 1
//                    hm.put(value, hm.getOrDefault(value, 0) + 1)
                    sb = StringBuilder()
                    flag = false
                }
            }
        }

        val sortedMap = hm.toSortedMap(Comparator{
            o1, o2 -> hm[o2]!!.compareTo(hm[o1]!!)
        }).keys.toIntArray()

        return sortedMap
    }
}

fun main(){
    val s = "{{2},{2,1},{2,1,3},{2,1,3,4}}"
    val sol = Solution()

    println(sol.solution(s).contentToString())
}